### ISSUE TITLE
Compilation fail with BOZ literal constant

### ISSUE THREAD

**Issue by davidpfister** [fortran-lang/fpm#1016]
### Description

I am trying to compile a code performing some SHA1 hashing. The code contains the following data
```fortran
data(h0_ref(i), i=1, 5) / &
            z'67452301', &
            z'EFCDAB89', &
            z'98BADCFE', &
            z'10325476', &
            z'C3D2E1F0'/
```
when compiling with fpm using `fpm build` I get the following message
```
Error: BOZ literal constant at (1) is neither a DATA statement value nor an actual argument of INT/REAL/DBLE/CMPLX intrinsic subprogram [see '-fno-allow-invalid-boz']
compilation terminated due to -fmax-errors=1.
<ERROR> Compilation failed for object " src_SHA1.f90.o "
<ERROR> stopping due to failed compilation
STOP 1
```

I manage to compile my code if I pass the proposed flag `option build --flag "-fallow-invalid-boz"`. However, there is a difference with ifort which compiles the same code without any problem and without specifying any additional flag. 

### Expected Behaviour

The code should compile with gfortran without passing additional flags just like it does with ifort

### Version of fpm

0.10

### Platform and Architecture

Windows

**Comment by ivan-pi** [comment]
Which version of gfortran were you using?

I don't see any problem starting with gfortran 10: https://godbolt.org/z/a7Wzrfj1x

Details around the usage of BOZ constants were standardized in recent Fortran versions, see https://stevelionel.com/drfortran/2020/08/11/doctor-fortran-in-were-all-bozos-on-this-bus/

**Comment by davidpfister** [comment]
@ivan-pi, I am using gfortran 14. And actually, gfortran can compile the code provided that I add the argument '-fallow-invalid-boz' (which may even be present by default). 
So there is no bug issue here, it's more that there is a discrepancy between compilers coming from default options that are used (or not) in fpm. fpm seems to enforce `-fno-allow-invalid-boz` for gfortran, but not the equivalent option (if it exists) for ifort.

**Comment by ivan-pi** [comment]
It looks like some handling was added for the MPI packages, but I don't see it being applied generally.

https://github.com/fortran-lang/fpm/blob/1ae7351a840e572043e954207ec7c30b03cb3521/src/fpm_meta.f90#L751-L761

I created a new fpm project, and added the following to `app/main.f90`

```
subroutine test
implicit none
integer :: h0_ref(5)
integer :: i

data (h0_ref(i), i=1, 5) / &
            z'67452301', &
            z'EFCDAB89', &
            z'98BADCFE', &
            z'10325476', &
            z'C3D2E1F0'/

print *, h0_ref

end subroutine

interface
  subroutine test
  end subroutine
end interface

call test
end
```

It builds without extra flags needed in fpm 0.11.0,

```
$ FPM_FC=gfortran fpm run
[  0%]                       main.f90
[ 50%]                       main.f90  done.
[ 50%]                            sm3
[100%]                            sm3  done.
[100%] Project compiled successfully.
  1732584193  -271733879 -1732584194   271733878 -1009589776
```

**Comment by ivan-pi** [comment]
> But the discrepancy between compilers remains. The implicit casting is not standard compliant I guess?

The interpretation of BOZ constants was clarified in F2008. You need to use,
```
        k = int(z'6ED9EBA1')
```
to make it standard-conforming, or add `-fallow-invalid-boz` in case of gfortran.

IMO this is not an issue fpm should fix.

**Comment by davidpfister** [comment]
I tend to agree. Seems that ifort is quite permissive with that. Is there an option to pass to ifort to enforce a behavior similar to gfortran and nag?

**Comment by davidpfister** [comment]
I guess it has not been [fixed](https://community.intel.com/t5/Intel-Fortran-Compiler/hexadecimal-numbers/td-p/1083892?profile.language=en&countrylabel=Latin%252525252525252520America) after all 😆 

![Image](https://github.com/user-attachments/assets/69cd24df-35a1-4f21-ae0f-313123bde680)
